### PR TITLE
Demonstrate augmented Lagrange preconditioning for Taylor Hood elements

### PR DESCRIPTION
**Pull request by lindsayad**
Refs #27126

### CODE REVIEW

File: modules/navier_stokes/doc/content/bib/navier_stokes.bib [hunk 1]
```diff
@@ -422,3 +422,34 @@ @article{zanetti2020scalable
...
+  journal={SIAM Journal on Scientific Computing},
+  volume={41},
+  number={5},
+  pages={A3073--A3096},
+  year={2019},
+  publisher={SIAM}
+}
+
+@article{benzi2023solving,
+  title={Solving linear systems of the form by preconditioned iterative methods},
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/bib/navier_stokes.bib:449]: title didnt paste well

File: modules/navier_stokes/include/base/NavierStokesMethods.h [hunk 2]
```diff
@@ -148,4 +148,22 @@ void getElementFaceArgs(const std::vector<BoundaryName> & wall_boundary_name,
...
+           const VectorType & value,
+           const PointType & point,
+           const Moose::CoordinateSystemType & coord_sys,
+           const unsigned int rz_radial_coord)
+{
+  auto div = gradient.tr();
+  if (coord_sys == Moose::COORD_RZ)
+    // u_r / r
+    div += value(rz_radial_coord) / point(rz_radial_coord);
+  return div;
```
Comment (GiudGiud) [inline modules/navier_stokes/include/base/NavierStokesMethods.h:167]: mooseAssert not spherical? or mooseAssert XYZ or RZ

File: modules/navier_stokes/include/kernels/INSADMomentumGradDiv.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+  INSADMomentumGradDiv(const InputParameters & parameters);
+
+protected:
+  ADReal computeQpResidual() override;
+
+  /// The grad-div stabilization coefficient
+  const Real _gamma;
+
+  const Moose::CoordinateSystemType & _coord_sys;
```
Comment (GiudGiud) [inline modules/navier_stokes/include/kernels/INSADMomentumGradDiv.h:30]:
```suggestion
  /// Coordinate system of the mesh
  const Moose::CoordinateSystemType & _coord_sys;
```

File: modules/navier_stokes/doc/content/source/kernels/INSADMomentumGradDiv.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+approximated by
+
+\begin{equation}
+S^{-1} = -\left(\nu + \gamma\right)M_p^{-1}
+\end{equation}
+
+where $\nu$ is the kinematic viscosity and $M_p$ is the pressure mass matrix
+([!citep](farrell2019augmented)). In [!citep](benzi2006augmented) the authors
+found that $\gamma$ on the order of the velocity magnitude provides $\nu$- and
+$h$-independent (the latter denoting element size) convergence for a wide range
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/source/kernels/INSADMomentumGradDiv.md:24]: so gamma is ok as a constant? or do they make it a field

**Review by GiudGiud — COMMENTED** [review]
looks pretty cool. I ll try to read one of the 3 to get up to speed

File: modules/navier_stokes/doc/content/source/kernels/INSADMomentumGradDiv.md (hunk 4 above)
Comment (lindsayad) [reply modules/navier_stokes/doc/content/source/kernels/INSADMomentumGradDiv.md:24]: it should be a constant

File: modules/navier_stokes/test/tests/finite_element/ins/lid_driven/steady_vector_fsp_al.i [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    type = INSADMomentumPressure
+    variable = vel
+    pressure = p
+    integrate_p_by_parts = true
+  []
+  [momentum_graddiv]
+    type = INSADMomentumGradDiv
+    variable = vel
+    gamma = ${gamma}
+  []
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_element/ins/lid_driven/steady_vector_fsp_al.i:64]: should this contribute to the residual too?
I guess I ll know once I read the paper but I was expecting it from the doco to only contribute to the mass matrix

File: modules/navier_stokes/doc/content/source/kernels/INSADMomentumGradDiv.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+!listing modules/navier_stokes/test/tests/finite_element/ins/lid_driven/steady_vector_fsp_al.i block=Kernels/mass_kernel
+
+The field split options are shown in
+
+!listing modules/navier_stokes/test/tests/finite_element/ins/lid_driven/steady_vector_fsp_al.i block=Preconditioning
+
+In this case we only wish to display the effectiveness of the pressure mass
+matrix as a preconditioner for the Schur complement, so we use LU solves for
+inverting the velocity block and $M_p$. Indeed for $\nu=10^{-3}$ (equivalent to
+$\textrm{Re}=1000$), the solution of the Schur complement system takes no more
```
Comment (grmnptr) [inline modules/navier_stokes/doc/content/source/kernels/INSADMomentumGradDiv.md:46]:
```suggestion
$\textrm{Re}=1000$), the solution of the Schur complement system to a relative tolerance of ($10^{-4}$) takes no more
```

File: modules/navier_stokes/test/tests/finite_element/ins/lid_driven/tests [hunk 7]
```diff
@@ -96,13 +96,23 @@ finite elements for velocity and pressure.'
...
+      exodiff = 'steady_vector_fsp_stokes_exo.e'
+      petsc_version = '>=3.20.0'
+      detail = 'a Stokes problem and a'
+    []
+    [high_Re]
+      type = 'Exodiff'
+      input = 'steady_vector_fsp_al.i'
+      exodiff = 'steady_vector_fsp_al_exo.e'
+      petsc_version = '>=3.20.0'
+      detail = 'a high Reynolds number (100) Navier-Stokes problem with a grad-div stabilization.'
```
Comment (grmnptr) [inline modules/navier_stokes/test/tests/finite_element/ins/lid_driven/tests:114]:
```suggestion
      detail = 'a high Reynolds number (1000) Navier-Stokes problem with a grad-div stabilization.'
```

**Review by grmnptr — CHANGES_REQUESTED** [review]
Nice new feature!

File: modules/navier_stokes/test/tests/finite_element/ins/lid_driven/steady_vector_fsp_al.i (hunk 5 above)
Comment (GiudGiud) [reply modules/navier_stokes/test/tests/finite_element/ins/lid_driven/steady_vector_fsp_al.i:64]: yup I see it equation 2.6 of Farrell

**Review by GiudGiud — APPROVED** [review]

**Review by grmnptr — APPROVED** [review]

**Comment by lindsayad** [comment]
failures are unrelated

**Comment by lindsayad** [comment]
thanks guys for the quick reviews!

**Pull request merged**